### PR TITLE
GuidedTours: (blanket) exit by guessing user intent

### LINKED ISSUE #5787: GuidedTours: dismiss tour automatically on other clicks on the page
Guided tours should stop automatically at any stage if the user instead of following it starts clicking on other active items on the page.

---

From comments on #5689: 

> Folletto: The tour doesn't dismiss if the user ignores it at some point and just clicks around somewhere else. It feels to me in that scenario the tour should stop itself, and let the user to get around freely.
 
> Shaun: I had the same experience a few times. It really easy to click somewhere else and still have the tour open — and it just doesn't make sense.

### PR DESCRIPTION
**Pull request by lsinger**
Sometimes, a user just wants to navigate away from a Guided Tour without necessarily clicking the correct "Exit Tour" button. Still, we'd like to quit in those cases where user intent is reasonably clear.

This change quits a currently running tour in the following cases:

- The step has a "target" element that it points to or expects an interaction with, but the user navigates to a new route where the target isn't available anymore.
- The step waits for the user the click "Next", but the user navigates to a different route.

This doesn't work in all cases. We'd rather have false negatives (trouble quitting a tour without clicking the "Exit Tour" button) than any false positives (unintended quits).

This closes #5787 and #5861. 

### To do

- [x] Consuming `sections.js` to fix `pathToSection`; also, that method could be a standalone function, preferably somewhere reusable
- [x] Use a cached version of `compose( last, getActionLog )`
- [x] **Rebase**
- ~~[ ] Analytics, the bare minimum to collect automated quits~~ (_in a subsequent PR_)

### To test

- go to http://calypso.localhost:3000/?tour=main or https://calypso.live/?branch=add/guided-tours-blanket-exit-only&tour=main
- click on "My Sites"
- ensure the tour doesn't appear again
- go back to http://calypso.localhost:3000/?tour=main or https://calypso.live/?branch=add/guided-tours-blanket-exit-only&tour=main (with a proper new load or reload)
- click through the tour until the step that says "This shows your currently selected site's name and address. Click your site's name to continue."
- click on "Reader"
- ensure the tour doesn't appear again
- go back to http://calypso.localhost:3000/?tour=main or https://calypso.live/?branch=add/guided-tours-blanket-exit-only&tour=main (with a proper new load or reload)
- go through the tour normally, following all the instructions
- ensure the tour steps all work and you get to the step that says "That's it! Now that you know a few of the basics, feel free to wander around." at the end

Test live: https://calypso.live/?branch=add/guided-tours-blanket-exit-only

### CODE REVIEW

**Comment by bisko** [comment]
I stumbled upon two problems with testing:

## Tour sometimes shows up in My Sites

When I do the following:

* Go to `http://calypso.localhost:3000/?tour=main` (tour shows)
* Click `My Sites`

Sometimes the tour pops up again while under `My Sites`.

I think this is best reproducible with a completely new account that is created in a clean incognito session. After some back and forth with the steps I described above, the tour stops showing up under `My Sites`

Video: http://d.pr/i/1bYoA

## Tour cannot be completed 

Sometimes the tour cannot be completed.

Steps to reproduce:

* Go to `http://calypso.localhost:3000/?tour=main` (tour shows)
* Follow the tour
* Click on `My Sites` when the tour instructs you to (step 2)
* You get sent to the Stats screen

Sometimes the tour disappears right after the page loads. I suspect a race condition with the First View popup. 

Video: http://d.pr/i/wKBC

File: client/layout/guided-tours/config-elements.js [hunk 1]
```diff
@@ -76,7 +80,7 @@ export class Tour extends Component {
 			return null;
 		}
 
-		return React.cloneElement( nextStep, { isLastStep } );
+		return React.cloneElement( nextStep, { isLastStep, isValid, lastAction } );
```
Comment (mcsf) [inline client/layout/guided-tours/config-elements.js]: Let's not pass these as props here, since they are made available to all descendants via context.

Comment (mcsf) [inline client/layout/guided-tours/config-elements.js]: (As done [here](https://github.com/Automattic/wp-calypso/pull/7431/files#diff-593a9c85e742030f6ecf82191238d1b5R97))

**Comment by lsinger** [comment]
I addressed @mcsf's comment in 481d51c. 

@bisko: I now can't reproduce the problem you've been seeing. Can you pull the latest from this branch, and then enable debug like so: `localStorage.debug = 'calypso:guided-tours'` Seeing that output would help me debug the issue if you can still reproduce it.

**Comment by lsinger** [comment]
> @bisko: I now can't reproduce the problem no. 1 you've been seeing. 

And now I also can't reproduce problem no. 2 anymore. Weird. I did `make clean distclean` at some point, but ... :puzzled:

**Comment by lsinger** [comment]
OK, I *can* reproduce problem no. 1. 

The — or at least *one* — problem seems to be: `this.section` and `lastAction.path` regularly point to the same thing. E.g. when I start on `/?tour=main` and then click `My Sites`, I'll have `/stats` for `this.section` and `/stats/...` for `lastAction.path`.

Is maybe `lastAction` assigned too early? Is it currently kind of the "current action" instead of the "previous action"? 

/cc @mcsf in case you remember what exactly we were thinking

**Comment by mcsf** [comment]
The changes I worked on in d5a42a2 got rid of the issue, but the actual committed changes don't. My hunch is that it's a synchronization issue, in which assigning `this.section` based on `location.pathname` happens too soon. `location` is impure, anyway, so maybe we can use actual `section` data from our state tree.

File: client/layout/guided-tours/config-elements.js [hunk 2]
```diff
@@ -116,19 +99,74 @@ export class Step extends Component {
...
 	}
 
+	setSection() {
+		// keep track of current section for "blanket exit"
+		// (i.e. quitIfInvalidRoute in `Step` and `Continue`)
+		// TODO(mcsf,lsinger): works only in a few cases and could be more elegant
+		// e.g. doesn't work when switching from reader start to discover
+		// as the Step gets re-created anew
+		// (better than false positives, though)
+		this.section = this.pathToSection( location.pathname );
```
Comment (mcsf) [inline client/layout/guided-tours/config-elements.js]: Fix this guy here, use proper state instead, probably adding `sectionName` or something to the context.

**Comment by mcsf** [comment]
Latest commits are a significant step to getting this done. Remaining tasks include:

- [ ] Consuming `sections.js` to fix `pathToSection`; also, that method could be a standalone function, preferably somewhere reusable
- [ ] Use a cached version of `compose( last, getActionLog )`
- [ ] ? … can't come up with anything else right now

**Comment by mcsf** [comment]
The only remaining item on the _to do_ list is analytics. I suggest we address that in a separate PR, once this PR and #7431 lands.

**Comment by mcsf** [comment]
@bisko: thank you for the testing! As of the latest commit, both issues should be fixed.

**Comment by marekhrabe** [comment]
I just tested this and I was able to follow test instructions without any problem. I also tried to reproduce problems @bisko described, without luck. 

Tested with my a8c account and also freshly created one.

**Comment by bisko** [comment]
I ran through the PR again and it seems to be working correctly! 🙇 

One small issue I noticed is when I open `?tour=main`, it sends me to the Reader with the start screen in the tour. If I click `My Sites` then, the tour takes a bit to disappear and it actually disappears when the stats screen has already started rendering. Is that an expected behavior? 

Here's a video of the problem: http://d.pr/v/1halv (pause the video and click around the 3rd or 4th second): 

screen shot 2016-10-18 at 13 47 57

It is possible that this is bug that is not related to this PR as I see that the switch from `Reader` to `My Sites` in the header happens a bit later too.

**Comment by lsinger** [comment]
@bisko thanks! We need to be a bit deliberate where we exit the tour as we don't want to exit prematurely, so we explicitly wait until `sectionIsLoading` is `false` again after it has been `true`. We pass that into Guided Tours as the `shouldPause` attribute.

File: client/layout/guided-tours/config-elements.js [hunk 3]
```diff
@@ -6,35 +8,52 @@ import classNames from 'classnames';
...
 import {
 	debounce,
+	defer,
 	find,
 	mapValues,
 	omit,
 	property,
+	startsWith,
 } from 'lodash';
+import debugFactory from 'debug';
```
Comment (lsinger) [inline client/layout/guided-tours/config-elements.js]: I talked to a few colleagues and the consensus seems to be to keep the more useful `debug` statements. I went through those we still have in here and think that all of them would be useful beyond the issue addressed in this PR — they all help us understand the life cycle of a tour. That will be helpful in the future. 

Also in production, these all become no-ops: https://github.com/Automattic/wp-calypso/blob/master/webpack.config.js#L147

TL;DR: yes.

**Comment by bisko** [comment]
Code looks good to me and it is working as expected. 👍
